rust/src/pgsql/log.rs

```diff
@@ -0,0 +1,420 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use std;
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+use crate::pgsql::pgsql::*;
+use crate::pgsql::parser::*;
+
+fn log_pgsql(tx: &PgsqlTransaction, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.set_uint("transaction_id", tx.tx_id.into())?;
+    if !tx.requests.is_empty() {
+        if tx.requests.len() > 1 {
+            js.open_array("requests")?;
+            for request in &tx.requests {
+                let jbr = log_request(&request)?;
+                js.append_object(&jbr)?;
+            }
+            js.close()?;
+        } else {
+            let jbr = log_request(&tx.requests[0])?;
+            js.set_object("request", &jbr)?;
+        }
+        if tx.responses.len() > 1 {
+            js.open_array("responses")?;
+            for response in &tx.responses {
+                let jb = log_response(&response)?;
+                js.append_object(&jb)?;
+            }
+            js.close()?;
+        }
+        else if !tx.responses.is_empty() {
+            let jb = log_response(&tx.responses[0])?;
+            js.set_object("response", &jb)?;
+        }
+    } else {
+        SCLogNotice!("Suricata logging a transaction without finding a request. Tx_id {:?}", &tx.tx_id);
+        if !tx.responses.is_empty() {
+            js.open_array("response")?;
+            for response in &tx.responses {
+                let jb = log_response(&response)?;
+                js.append_object(&jb)?;
+            }
+            js.close()?;
+        } else {
+            js.set_string("request", "request is empty")?;
+            js.set_string("response", "response is empty")?;
+            SCLogNotice!("Warning, Suricata created an empty PGSQL transaction");
+        }
+    }
+    Ok(())
+}
+
+fn log_request(req: &PgsqlFEMessage) -> Result<JsonBuilder, JsonError>
+{
+    let mut js = JsonBuilder::new_object();
+    js.open_object(&req.get_message_type())?;
+    match req {
+        PgsqlFEMessage::StartupMessage(
+            StartupPacket{
+                length: _,
+                proto_major,
+                proto_minor,
+                params}) =>
+        {
+            let proto = format!("{}.{}", proto_major, proto_minor);
+            js.set_string("protocol_version", &proto)?;
+            let jb = log_pgsql_parameters(params)?;
+            js.set_object("startup_parameters", &jb)?;
+        },
+        PgsqlFEMessage::SslRequest(_) =>
+        {
+            js.set_string("message_type", "SSL Request")?;
```

Comment:
I think this should be `SSLRequest` as it seems to be consistently like that in the Postgres docs.